AWSPinpoint/AWSPinpointEndpointProfile.m

```diff
@@ -546,4 +561,83 @@ - (void)encodeWithCoder:(NSCoder *)encoder {
     [encoder encodeObject:_userId forKey:@"userId"];
 }
 
+- (void)addUserAttribute:(NSArray *)theValue
+                  forKey:(NSString *)theKey {
+    if (!theKey) return;
+    @synchronized(self) {
+        if (theKey != nil && theValue != nil && ![self hasUserAttributeForKey:theKey]) {
+            if (self.currentNumOfUserAttributes < MAX_NUM_OF_METRICS_AND_ATTRIBUTES) {
+                [self.userAttributes setValue:[AWSPinpointEndpointProfileUser processUserAttributeValues:theValue]
+                                       forKey:[AWSPinpointEndpointProfileUser trimKey:theKey
+                                                                              forType:@"userAttribute"]];
+            } else {
+                AWSDDLogWarn(@"Max number of userAttributes/metrics reached, dropping attribute with key: %@", theKey);
+            }
+        } else if ([self hasUserAttributeForKey:theKey]) {
+            [self.userAttributes setValue:[AWSPinpointEndpointProfileUser processUserAttributeValues:theValue]
+                                   forKey:[AWSPinpointEndpointProfileUser trimKey:theKey
+                                                                          forType:@"userAttribute"]];
+        }
+    }
+}
+
+- (NSDictionary *)allUserAttributes {
+    @synchronized(self) {
+        return [NSDictionary dictionaryWithDictionary:self.userAttributes];
+    }
+}
+
+- (BOOL)hasUserAttributeForKey:(NSString *)theKey {
+    if(!theKey) return NO;
+    @synchronized(self) {
+        if([self.userAttributes objectForKey:theKey]) {
+            return YES;
+        } else {
+            return NO;
+        }
+    }
+}
+
+- (NSArray *)userAttributeForKey:(NSString *)theKey {
+    @synchronized(self) {
+        return [self.userAttributes objectForKey:theKey];
+    }
+}
+
++ (NSArray *)processUserAttributeValues:(NSArray*) values {
+    NSMutableArray *trimmedValues = [NSMutableArray arrayWithCapacity:MAX_ENDPOINT_ATTRIBUTE_VALUES];
+    int valuesCount = 0;
+    for (NSString *val in values) {
+        [trimmedValues addObject:[AWSPinpointEndpointProfileUser trimValue:val]];
+        if (++valuesCount >= MAX_ENDPOINT_ATTRIBUTE_VALUES) {
+            AWSDDLogWarn(@"Only %d attributes values are allowed, attribute values has been reduced to %d values.", MAX_ENDPOINT_ATTRIBUTE_VALUES, MAX_ENDPOINT_ATTRIBUTE_VALUES);
+            break;
+        }
+    }
+    return trimmedValues;
```

Comment:
I think we should come up with a different algorithm.  This will lead to non-deterministic behavior since the order or values can vary.

Reply:
I am hesitant to put any of this error checking/boundary logic into this code (that we would need to maintain).  I understand that we already have code like this in our codebase, but I am against continuing to add additional logic in our SDK that has this type of restricting logic, because I think the service should dictate what values it is allowed to accept -- not the client.  I actually tried adding 100 items to a single attribute and i got back the following error:

```
2020-10-24 15:01:54:733 SDKIssue3086[58900:645291] Response body:
{"Message":"Too many attribute values for CustomUserAttribute"}
2020-10-24 15:01:54:733 SDKIssue3086[58900:645291] Unable to successfully update endpoint. Error Message:Error Domain=com.amazonaws.AWSPinpointTargetingErrorDomain Code=1 "BadRequestException" UserInfo={responseStatusCode=400, responseHeaders=<CFBasicHash 0x600002a54440 [0x7fff8062d750]>{type = immutable dict, count = 12,
entries =>
```